zipkin-storage/cassandra3/src/main/java/zipkin/storage/cassandra3/CassandraUtil.java

```diff
@@ -73,19 +73,20 @@ public static int durationIndexBucket(long ts_micro) {
       if (Constants.CORE_ANNOTATIONS.contains(a.value)) continue;
 
       if (a.endpoint != null && !a.endpoint.serviceName.isEmpty()) {
-        annotationKeys.add(a.endpoint.serviceName + ":" + a.value);
+        annotationKeys.add(a.endpoint.serviceName);
+        annotationKeys.add(a.value);
       }
     }
     for (BinaryAnnotation b : span.binaryAnnotations) {
-      if (b.type == BinaryAnnotation.Type.STRING
-          && b.endpoint != null
-          && !b.endpoint.serviceName.isEmpty()
-          && b.value.length <= LONGEST_VALUE_TO_INDEX * 4) { // UTF_8 is up to 4bytes/char
-        String value = new String(b.value, UTF_8);
-        if (value.length() > LONGEST_VALUE_TO_INDEX) continue;
-
-        annotationKeys.add(b.endpoint.serviceName + ":" + b.key);
-        annotationKeys.add(b.endpoint.serviceName + ":" + b.key + ":" + new String(b.value, UTF_8));
+      if (b.endpoint != null && !b.endpoint.serviceName.isEmpty() && !Constants.CORE_ANNOTATIONS.contains(b.key)) {
+        annotationKeys.add(b.endpoint.serviceName);
+        if (b.type == BinaryAnnotation.Type.STRING) {
+            String value = new String(b.value, UTF_8);
+            value = value.substring(0, Math.min(value.length(), LONGEST_VALUE_TO_INDEX));
+            annotationKeys.add(b.key + ":" + value);
+        } else {
+            annotationKeys.add(b.key);
```

Comment:
but we can…
i've tested this and it means that searching on non-string binary annotations is possible.

feasibility of it aside, has it value?